presto-main/src/test/java/com/facebook/presto/type/TestRowOperators.java

```diff
@@ -124,6 +151,161 @@ public void testRowToJson()
+    {
+        // special values
+        assertFunction("CAST(CAST (null AS JSON) AS ROW(BIGINT))", new RowType(ImmutableList.of(BIGINT), Optional.empty()), null);
+        assertFunction("CAST(JSON 'null' AS ROW(BIGINT))", new RowType(ImmutableList.of(BIGINT), Optional.empty()), null);
+        assertFunction("CAST(JSON '[null, null]' AS ROW(VARCHAR, BIGINT))", new RowType(ImmutableList.of(VARCHAR, BIGINT), Optional.empty()), Lists.newArrayList(null, null));
+
+        // allow json object contains non-exist field name
+        assertFunction(
+                "CAST(JSON '{\"k1\": [1, 2], \"used\": 3, \"k2\": [4, 5]}' AS ROW(used BIGINT))",
+                new RowType(ImmutableList.of(BIGINT), Optional.of(ImmutableList.of("used"))),
+                ImmutableList.of(3L));
+        assertFunction(
+                "CAST(JSON '[{\"k1\": [1, 2], \"used\": 3, \"k2\": [4, 5]}]' AS ARRAY<ROW(used BIGINT)>)",
+                new ArrayType(new RowType(ImmutableList.of(BIGINT), Optional.of(ImmutableList.of("used")))),
+                ImmutableList.of(ImmutableList.of(3L)));
+
+        // fields out of order
+        assertFunction(
+                "CAST(unchecked_to_json('{\"k4\": 4, \"k2\": 2, \"k3\": 3, \"k1\": 1}') AS ROW(k1 BIGINT, k2 BIGINT, k3 BIGINT, k4 BIGINT))",
+                new RowType(ImmutableList.of(BIGINT, BIGINT, BIGINT, BIGINT), Optional.of(ImmutableList.of("k1", "k2", "k3", "k4"))),
+                ImmutableList.of(1L, 2L, 3L, 4L));
+        assertFunction(
+                "CAST(unchecked_to_json('[{\"k4\": 4, \"k2\": 2, \"k3\": 3, \"k1\": 1}]') AS ARRAY<ROW(k1 BIGINT, k2 BIGINT, k3 BIGINT, k4 BIGINT)>)",
+                new ArrayType(new RowType(ImmutableList.of(BIGINT, BIGINT, BIGINT, BIGINT), Optional.of(ImmutableList.of("k1", "k2", "k3", "k4")))),
+                ImmutableList.of(ImmutableList.of(1L, 2L, 3L, 4L)));
+
+        // boolean
+        assertFunction("CAST(JSON '[true, false, 12, 0, 12.3, 0.0, \"true\", \"false\", null]' AS ROW(BOOLEAN, BOOLEAN, BOOLEAN, BOOLEAN, BOOLEAN, BOOLEAN, BOOLEAN, BOOLEAN, BOOLEAN))",
+                new RowType(ImmutableList.of(BOOLEAN, BOOLEAN, BOOLEAN, BOOLEAN, BOOLEAN, BOOLEAN, BOOLEAN, BOOLEAN, BOOLEAN), Optional.empty()),
+                asList(true, false, true, false, true, false, true, false, null));
+
+        assertFunction("CAST(JSON '{\"k1\": true, \"k2\": false, \"k3\": 12, \"k4\": 0, \"k5\": 12.3, \"k6\": 0.0, \"k7\": \"true\", \"k8\": \"false\", \"k9\": null}' AS ROW(k1 BOOLEAN, k2 BOOLEAN, k3 BOOLEAN, k4 BOOLEAN, k5 BOOLEAN, k6 BOOLEAN, k7 BOOLEAN, k8 BOOLEAN, k9 BOOLEAN))",
+                new RowType(ImmutableList.of(BOOLEAN, BOOLEAN, BOOLEAN, BOOLEAN, BOOLEAN, BOOLEAN, BOOLEAN, BOOLEAN, BOOLEAN),
+                        Optional.of(ImmutableList.of("k1", "k2", "k3", "k4", "k5", "k6", "k7", "k8", "k9"))),
+                asList(true, false, true, false, true, false, true, false, null));
+
+        // tinyint, smallint, integer, bigint
+        assertFunction(
```

Comment:
Note: These tests are not as comprehensive as JSON -> ARRAY/MAP . Especially, it doesn't assert the behavior of different types of JsonToken cast. But I feel it's not necessary now to test it here.